feat(project): add stylelint-plugin-carbon-tokens package (#6799)
carbon-design-system/carbon: packages/stylelint-plugin-carbon-tokens/README.md

# stylelint-plugin-carbon-tokens

> A stylelint plugin to support the use of carbon component tokens.

## Getting started

To install `stylelint-plugin-carbon-tokens` in your project, you will need to
run the following command using [npm](https://www.npmjs.com/):

```bash

> npm install -S stylelint-plugin-carbon-tokens

Reviewer: this flag is redundant now isn't it? or should this be a dev dependency?

Author: Replaced with -D

```

If you prefer [Yarn](https://yarnpkg.com/en/), use the following command
instead:

```bash
yarn add stylelint-plugin-carbon-tokens
```

## Usage

This project is intended to help users identify cases where tokens, functions
and mixins defined as part of various projects relating to the Carbon Design

> System are used. [www.carbondesignsystem.com]

Reviewer: is this supposed to be a raw link or should it have a title?

Author: Changed to
```
see [www.carbondesignsystem.com](www.carbondesignsystem.com) 
```

Oddly we are not very self-referential in source, can change to something more appropriate if desired.

It consists of a stylelint plugin and depends on various Carbon Design System
packages for loading settings.

It includes, but may not be limited to, linting for @carbon/themes,
@carbon/colors, @carbon/layout, @carbon/type and @carbon/motion.

> Not incluced, as they're not used through SCSS, are Carbon Icons, Grid and any

Reviewer: ```suggestion
Not included, as they're not used through SCSS, are Carbon Icons, Grid and any
```

Author: Corrected spelling and installed spell checker extension. Do you have a favourite spelling extension for VSCode?

Reviewer 2: I've been using https://github.com/streetsidesoftware/vscode-spell-checker

other DOM related checks..

### Please be helpful

> Before we start this project is a work in progress which deliberately outputs

Reviewer: ```suggestion
Before we start, this project is a work in progress which deliberately outputs
```

Author: Corrected grammar.

> warnings when it comes accross a syntax that has not yet been catered for. If

Reviewer: ```suggestion
warnings when it comes across syntax that has not yet been catered for. If
```

Author: Corrected spelling and grammar, wonder if Grammarly has a VSCode plugin ;-)

        acceptCarbonTypeScaleFunction: false,
      },
    ],
    //...other rules
  },
  //...
};
```

NOTE: By default rules accept SCSS and CSS variables not defined in the current
file prior to their use. Set acceptUndefinedVariables to false to disable this
behaviour.

### Advanced options

These options when omitted to accept the defaults. They are intended to support
non-standard use cases and accept values that use a syntax which may well need
some refining as the project moves forward.

- includeProps: Array
- acceptValues: Array

Arrays of strings and/or Regex followed by a range in angled brackets.

The defaults for these are defined in the individual README files listed above.

- `includeProps: []` - Indicates default, same as omitting the property
- `includeProps: ["*"]` - Indicates default, same as omitting the property
- `includeProps: ["/^\\$my-color--/", "*"]` - SCSS variable starting
  "\\\$my-color--", plus default values specified

The last option here shows how you could elect to check your own tokens refer to
values acceptable to the linter.

- `acceptValues: ["$/^\\$my-color--/"]` - Accept SCSS variables starting
  "\\\$my-color--"

#### includeProps Range

> Can innclude a range value expressed inside greater than and less than signs.

Reviewer: ```suggestion
Can include a range value expressed inside greater than and less than signs.
```

Author: Corrected

e.g. `["/prop-a$/<-1>", "/prop-b$/<1 -2>"]`

The above specifies the last value of `prop-a` and the first to second last of
`prop-b`,

It can be applied either to regex or string values and allows values or a range
of values to be selected from a multipart value.

The range value allows values to be selected from a multipart value such as a
box-shadow.

- Positive values represent positions at the start of a value list e.g. 1 is the
  first value.

- Negative values represent positions at the end of a value list. e.g. -1 = last
  value

- If no range is specified the whole value list is checked.

- A single value means only that value in a list is checked

- Two values represent start and end values of a range in the list.

#### includeProps specific values

> For some props e.g. transform we are only intereste in values that match a

Reviewer: ```suggestion
For some props e.g. transform we are only interested in values that match a
```

Author: Corrected

certain criteria.

This is specified as part of the includedProp inside \[].

e.g. `translate[/^transform/]`

> In this case only values starting `transform` are tested so not `skew` for

Reviewer: ```suggestion
In this case only values starting with `transform` are tested, so not `skew` for
```
